[TextFields] Adding font themer

mohammadcazig opened the pull request:
FontThemer Support for Textfields.
Close #3095

mohammadcazig committed:
Extracting Accessibility out of FeatureHighlightViewController to a mutator.
Tests added.
Example updated.

mohammadcazig committed:
Revert "Extracting Accessibility out of FeatureHighlightViewController to a mutator."

This reverts commit 0b5433ebf86366087b240c4dc4992a0603fa73e0.

mohammadcazig committed:
Merge branch 'develop' of github.com:material-components/material-components-ios into develop

mohammadcazig committed:
Merge branch 'develop' of github.com:material-components/material-components-ios into develop

mohammadcazig committed:
Adding Font themer to TextFields.

mohammadcazig committed:
Merge branch 'develop' of github.com:material-components/material-components-ios into TextFields_Font_Themer

mohammadcazig committed:
extra lf.

mohammadcazig committed:
dealing with warnings.

mohammadcazig committed:
Fixing unit tests.

ianegordon commented:
kokoro/bazel is upset:
error: -[components_TextFields_unit_test_swift_sources.MultilineTextFieldTests testSerialization] : failed: caught "NSInternalInconsistencyException", "NSLayoutConstraint constant is not finite!  That's illegal.  constant:nan firstAnchor:<NSLayoutYAxisAnchor:0x7fb246567a60 "MDCIntrinsicHeightTextView:0x7fb245808600'Lorem ipsum dolor sit ame...'.top"> secondAnchor:<NSLayoutYAxisAnchor:0x7fb243716a60 "MDCMultilineTextField:0x7fb243705600.top">"
(
	0   CoreFoundation                      0x000000010c8101ab __exceptionPreprocess + 171
	1   libobjc.A.dylib                     0x000000010bea5f41 objc_exception_throw + 48
	2   CoreFoundation                      0x000000010c815372 +[NSException raise:format:arguments:] + 98
	3   Foundation                          0x000000010b819089 -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:] + 193
	4   Foundation                          0x000000010b79d28c -[NSLayoutConstraint _setSymbolicConstant:constant:] + 629
	5   Foundation                          0x000000010b79cfe9 -[NSLayoutConstraint setConstant:] + 68
	6   Foundation                          0x000000010b79cce2 +[NSLayoutConstraint constraintWithItem:attribute:relatedBy:toItem:attribute:multiplier:constant:] + 293
	7   unit_tests                          0x0000000122682320 -[MDCMultilineTextField updateConstraints] + 1296
	8   UIKit                               0x0000000110125c10 -[UIView(AdditionalLayoutSupport) _sendUpdateConstraintsIfNecessaryForSecondPass:] + 161
	9   UIKit                               0x000000011012622c -[UIView(AdditionalLayoutSupport) _updateConstraintsIfNeededCollectingViews:forSecondPass:] + 1296
	10  Foundation                          0x000000010b7a3de1 -[NSISEngine withBehaviors:performModifications:] + 131
	11  UIKit                               0x0000000110126a5d __100-[UIView(AdditionalLayoutSupport) _updateConstraintsIfNeededWithViewForVariableChangeNotifications:]_block_invoke + 90
	12  UIKit                               0x00000001101252bb -[UIView(AdditionalLayoutSupport) _withUnsatisfiableConstraintsLoggingSuspendedIfEngineDelegateExists:] + 104
	13  UIKit                               0x00000001101265cc -[UIView(AdditionalLayoutSupport) _updateConstraintsIfNeededWithViewForVariableChangeNotifications:] + 160
	14  UIKit                               0x00000001101276e6 -[UIView(AdditionalLayoutSupport) _updateConstraintsAtEngineLevelIfNeededWithViewForVariableChangeNotifications:] + 401
	15  UIKit                               0x000000010f66ca8c -[UIView(Hierarchy) layoutBelowIfNeeded] + 1517
	16  unit_tests                          0x000000012268f0b8 -[MDCTextInputControllerBase movePlaceholderToUp:] + 312
	17  unit_tests                          0x00000001226904f5 -[MDCTextInputControllerBase forceUpdatePlaceholderY] + 341
	18  unit_tests                          0x000000012268dd6a -[MDCTextInputControllerBase setupInput] + 938
	19  unit_tests                          0x000000012268c373 -[MDCTextInputControllerBase initWithTextInput:] + 147
	20  unit_tests                          0x000000012262a8b2 _TTOFCSo31MDCTextInputControllerUnderlinecfT9textInputGSqCSo6UIView__S_ + 34
	21  unit_tests                          0x00000001226166e1 _TFCSo31MDCTextInputControllerUnderlineCfT9textInputGSqCSo6UIView__S_ + 65
	22  unit_tests                          0x00000001225e8b71 _TFC45components_TextFields_unit_test_swift_sources23MultilineTextFieldTests17testSerializationfT_T_ + 449
	23  unit_tests                          0x00000001225ec9c2 _TToFC45components_TextFields_unit_test_swift_sources23MultilineTextFieldTests17testSerializationfT_T_ + 34
	24  CoreFoundation                      0x000000010c79449c __invoking___ + 140
	25  CoreFoundation                      0x000000010c794370 -[NSInvocation invoke] + 320
	26  XCTest                              0x000000010bd8d8e2 __24-[XCTestCase invokeTest]_block_invoke_2 + 491
	27  XCTest                              0x000000010bdcbf62 -[XCTestContext performInScope:] + 190
	28  XCTest                              0x000000010bd8d6e4 -[XCTestCase invokeTest] + 254
	29  XCTest                              0x000000010bd8dfb2 -[XCTestCase performTest:] + 565
	30  XCTest                              0x000000010bd8b078 __27-[XCTestSuite performTest:]_block_invoke + 300
	31  XCTest                              0x000000010bd8ac68 -[XCTestSuite _performProtectedSectionForTest:testSection:] + 29
	32  XCTest                              0x000000010bd8ae4e -[XCTestSuite performTest:] + 214
	33  XCTest                              0x000000010bd8b078 __27-[XCTestSuite performTest:]_block_invoke + 300
	34  XCTest                              0x000000010bd8ac68 -[XCTestSuite _performProtectedSectionForTest:testSection:] + 29
	35  XCTest                              0x000000010bd8ae4e -[XCTestSuite performTest:] + 214
	36  XCTest                              0x000000010bd8b078 __27-[XCTestSuite performTest:]_block_invoke + 300
	37  XCTest                              0x000000010bd8ac68 -[XCTestSuite _performProtectedSectionForTest:testSection:] + 29
	38  XCTest                              0x000000010bd8ae4e -[XCTestSuite performTest:] + 214
	39  XCTest                              0x000000010bdd7a86 __44-[XCTTestRunSession runTestsAndReturnError:]_block_invoke + 40
	40  XCTest                              0x000000010bd99f27 -[XCTestObservationCenter _observeTestExecutionForBlock:] + 574
	41  XCTest                              0x000000010bdd7925 -[XCTTestRunSession runTestsAndReturnError:] + 281
	42  XCTest                              0x000000010bd75a50 -[XCTestDriver runTestsAndReturnError:] + 254
	43  XCTest                              0x000000010bdcf86e _XCTestMain + 559
	44  xctest                              0x000000010b6c4679 xctest + 9849
	45  libdyld.dylib                       0x000000010dfecd81 start + 1
)

Can you repro locally?

mohammadcazig commented:
@ianegordon  I was not able to repro this. I didn't run blaze though.

mohammadcazig committed:
commenting out low priority set up for bounds

mohammadcazig committed:
commenting out tests to see if tests would fail.

mohammadcazig committed:
tests are back, scale could have been zero.

mohammadcazig committed:
temp fix in case the fonts are zero.

mohammadcazig committed:
check for default inside setter.

mohammadcazig committed:
fix compliing mistakes.

mohammadcazig committed:
add documentation for the change.

mohammadcazig committed:
Merge branch 'develop' of github.com:material-components/material-components-ios into develop

mohammadcazig committed:
Example of dynamic size with different fonts.

ianegordon reviewed (dismissed):
Please add the testing sample per our offline discussion.

mohammadcazig committed:
Merge remote-tracking branch 'origin/TextFields_Dynamic_Size_Example' into TextFields_Font_Themer

mohammadcazig committed:
Merge branch 'develop' of github.com:material-components/material-components-ios into TextFields_Font_Themer

mohammadcazig committed:
added support for dynamic type with custom fonts.

mohammadcazig committed:
fixed the top overlapping issue

willlarche commented on components/TextFields/examples/TextFieldDynamicSizeExample.m:
```diff
@@ -0,0 +1,410 @@
+
+#import "MaterialTextFields.h"
+
+@interface TextFieldDynamicSizeExample : UIViewController <UITextFieldDelegate, UITextViewDelegate>
```
I don't love naming this `Dynamic Size Example`. From the user's point of view, it's really a `Custom Fonts Example`.

mohammadcazig replied:
Done.

willlarche commented on components/TextFields/examples/TextFieldDynamicSizeExample.m:
```diff
@@ -0,0 +1,410 @@
+
+@property(nonatomic) UIScrollView *scrollView;
+
+@property(nonatomic) UIFont *normalTextFont;
```
`UIFont` above `UIScrollView` in alphabet.

mohammadcazig replied:
Done.

willlarche commented on components/TextFields/examples/TextFieldDynamicSizeExample.m:
```diff
@@ -0,0 +1,410 @@
+
+  [self setupScrollView];
+
+  // System Font TextField
```
Too much going on in the `viewDidLoad`. Extract into at least 2 more (singleline setup and multiline setup).

mohammadcazig replied:
Done.

willlarche replied:
That's great. Gives you free documentation.

willlarche reviewed (changes requested):
It's so good!

willlarche commented on components/TextFields/examples/TextFieldDynamicSizeExample.m:
```diff
@@ -0,0 +1,410 @@
+  self.systemFontController.placeholderText = @"System Font";
+  self.systemFontController.mdc_adjustsFontForContentSizeCategory = NO;
+
+  // System Font Dynamic TextField
```
Expand these comments into semi-sentences.

willlarche replied:
I should have been clearer. I meant around the use of `Dynamic` and `System`. Sometimes you say `System font` which is clear. But sometimes you don't.

willlarche commented on components/TextFields/examples/TextFieldDynamicSizeExample.m:
```diff
@@ -0,0 +1,410 @@
+  NSMutableArray<NSLayoutConstraint *> *constraints = [NSMutableArray
+                                                       arrayWithArray:
+                                                       [NSLayoutConstraint
+                                                        constraintsWithVisualFormat:@"V:[t1]-[t2]-[t3]-[t4]-[t5]-[t6]-[t7]"
```
Add `-` before `[t1]`.

mohammadcazig replied:
t1 to  super view is being added in next lines.

willlarche replied:
Oh yeah! I forgot we had to do that.

willlarche commented on components/TextFields/examples/TextFieldDynamicSizeExample.m:
```diff
@@ -0,0 +1,410 @@
+- (BOOL)textFieldShouldReturn:(UITextField *)textField {
+  [textField resignFirstResponder];
+
+//  if (textField == (UITextField *)self.phoneController.textInput &&
```
Leave a comment for your comment.

mohammadcazig replied:
Oops this shouldn't be there. Got rid of it.

willlarche commented on components/TextFields/src/FontThemer/MDCTextFieldFontThemer.h:
```diff
@@ -0,0 +1,57 @@
+ @param textInputController a MDCTextInputController instance to apply a font scheme.
+ */
++ (void)applyFontScheme:(nonnull id<MDCFontScheme>)fontScheme
+  toTextInputController:(nullable id<MDCTextInputController>)textInputController;
```
The controller can be null?

mohammadcazig replied:
so technically there is no hurt in passing in nil in here. @romoore suggested we could make these nullable and I have no argument against it.

willlarche replied:
Even if there is no hurt, it will be a better experience in swift. Right now it takes an optional.

romoore replied:
I adopted this recommendation because many of our themers/mutators permit a `nullable` target and a `nonnull` color scheme. My thinking about it is that our API should define restrictions via nullability.  So if a method *can* safely accept a nil value, it should. If it cannot safely accept a nil value, it should restrict that parameter to `nonnull`.  Is it bad to accept optionals in Swift? I thought `nullable` has a broader scope than `nonnull`?

willlarche replied:
I'm not sure. Getting another opinion.

willlarche commented on components/TextFields/src/FontThemer/MDCTextFieldFontThemer.h:
```diff
@@ -0,0 +1,57 @@
+ Applies a font scheme to theme to a MDCTextInputController instance.
+
+ @param fontScheme The font scheme to apply to MDCTextInputController.
+ @param textInputController a MDCTextInputController instance to apply a font scheme.
```
`a` should be `A` if `The` is the pattern you want.

Add `a` between `to` and `MDCTextInputController`.

mohammadcazig replied:
Done

willlarche commented on components/TextFields/src/FontThemer/MDCTextFieldFontThemer.h:
```diff
@@ -0,0 +1,57 @@
+ @param textInputControllerClass a MDCTextInputController class to apply the font scheme to.
+ */
++ (void)applyFontScheme:(nonnull id<MDCFontScheme>)fontScheme
+  toAllTextInputControllersOfClass:(nullable Class<MDCTextInputController>)textInputControllerClass
```
Nullable?

mohammadcazig replied:
Replied to the first comment.

willlarche commented on components/TextFields/src/FontThemer/MDCTextFieldFontThemer.h:
```diff
@@ -0,0 +1,57 @@
+  toTextInputController:(nullable id<MDCTextInputController>)textInputController;
+
+/**
+ Applies a font scheme to theme to MDCTextInputController for a specific class type.
```
Does this apply it to existing instances of the class type or just future ones?

mohammadcazig replied:
I added extra comment, this would not apply to existing instances.

willlarche commented on components/TextFields/src/FontThemer/MDCTextFieldFontThemer.h:
```diff
@@ -0,0 +1,57 @@
+ Applies a font scheme to theme to MDCTextInputController for a specific class type.
+
+ @param fontScheme The font scheme to apply to MDCTextInputController.
+ @param textInputControllerClass a MDCTextInputController class to apply the font scheme to.
```
`a` becomes `A`.

mohammadcazig replied:
Done.

willlarche commented on components/TextFields/src/FontThemer/MDCTextFieldFontThemer.h:
```diff
@@ -0,0 +1,57 @@
+ */
++ (void)applyFontScheme:(nonnull id<MDCFontScheme>)fontScheme
+  toAllTextInputControllersOfClass:(nullable Class<MDCTextInputController>)textInputControllerClass
+  NS_SWIFT_NAME(apply(_:toAllControllersOfClass:));
```
Good.

willlarche commented on components/TextFields/src/FontThemer/MDCTextFieldFontThemer.h:
```diff
@@ -0,0 +1,57 @@
+  NS_SWIFT_NAME(apply(_:toAllControllersOfClass:));
+
+/**
+ Applies a font scheme to theme to MDCTextField instance.
```
Grammar not expressing what you intended.

mohammadcazig replied:
can you elaborate a bit more?

willlarche commented on components/TextFields/src/FontThemer/MDCTextFieldFontThemer.h:
```diff
@@ -0,0 +1,57 @@
+ Applies a font scheme to theme to MDCTextField instance.
+
+ @param fontScheme The font scheme to apply to MDCTextField.
+ @param textField a MDCTextField instance to apply the font scheme to.
```
`a` to `A`.

willlarche commented on components/TextFields/src/FontThemer/MDCTextFieldFontThemer.h:
```diff
@@ -0,0 +1,57 @@
+ @param textField a MDCTextField instance to apply the font scheme to.
+ */
++ (void)applyFontScheme:(nonnull id<MDCFontScheme>)fontScheme
+            toTextField:(nullable MDCTextField *)textField;
```
Nullable?

mohammadcazig replied:
Replied to the first comment.

willlarche commented on components/TextFields/src/FontThemer/MDCTextFieldFontThemer.m:
```diff
@@ -0,0 +1,75 @@
+  textField.font = fontScheme.body1;
+}
+
+// TODO: (larche) Drop this if defined and the pragmas when we drop Xcode 8 support.
```
I didn't write a perfectly clear comment. This would be better: `Drop this "#if !defined"`...

mohammadcazig replied:
Done

willlarche replied:
Much better. My personal thanks.

willlarche commented on components/TextFields/src/FontThemer/MDCTextFieldFontThemer.m:
```diff
@@ -0,0 +1,75 @@
+
++ (void)applyFontScheme:(id<MDCFontScheme>)fontScheme
+            toTextField:(MDCTextField *)textField {
+  textField.font = fontScheme.body1;
```
`MDCTextFields` have labels. Should you apply to them too?

mohammadcazig replied:
Added.

willlarche commented on components/TextFields/src/MDCTextInputControllerBase.m:
```diff
@@ -1068,6 +1068,10 @@ + (CGFloat)floatingPlaceholderScaleDefault {
 
 + (void)setFloatingPlaceholderScaleDefault:(CGFloat)floatingPlaceholderScaleDefault {
   _floatingPlaceholderScaleDefault = floatingPlaceholderScaleDefault;
+  if (floatingPlaceholderScaleDefault <= 0) {
```
Just curious, what was the situation where you found this was an issue?

mohammadcazig replied:
So for other properties we have null_resettable where we could set the property to null to reset it to default. but for this property that didn't exist and setting the value to zero or less is meaning less so I added the option. I also needed that incase we are applying a new font scheme which doesn't have enough info to set a new scale.

willlarche replied:
Thinking ahead! Got it.

willlarche reviewed (dismissed):
PTAL.

mohammadcazig committed:
Merge branch 'develop' of https://github.com/material-components/material-components-ios into TextFields_Font_Themer

mohammadcazig committed:
Replied to comment and fixed the issues.

mohammadcazig committed:
Adding Text Input property to InputController
Changing the value of mdc_adjustsFontForContentSizeCategory should not effect the actual controller fonts that are set by user but only the properties of the actual textfield. In order to do that I moved the resizing of fonts away from the properties of InputController and into update layout.

willlarche commented on components/TextFields/examples/TextFieldCustomFontExample.m:
```diff
@@ -393,18 +426,18 @@ - (BOOL)isValidPhoneNumber:(NSString *)inputString partially:(BOOL)isPartialChec
 
 + (BOOL)catalogIsPresentable {
-  return YES;
+  return NO;
```
What does this do again?

mohammadcazig replied:
It's only shown in Dragon but not Catalog.

willlarche commented on components/TextFields/src/FontThemer/MDCTextFieldFontThemer.h:
```diff
@@ -0,0 +1,58 @@
+  toTextInputController:(nullable id<MDCTextInputController>)textInputController;
+
+/**
+ Applies a font scheme to theme to MDCTextInputController for a specific class type. Would not apply
```
`Will not apply` would sound a little better.

willlarche replied:
`Applies a font scheme to theme to MDCTextInputController` is an unclear phrase. That's what I meant in the comment from before.

mohammadcazig replied:
I changed the phrasing.

willlarche reviewed (dismissed):
Nullable thing I'm not convinced on. Let me know IRL if you wanna talk it out more. That and one comment that needs clarification.

mohammadcazig committed:
Fixing the comment. Changing font scheme to not accept nullable controllers.

mohammadcazig committed:
change documentation.
